File: parsers/src/main/java/vct/parsers/ColPVLParser.java
package vct.parsers;

import hre.lang.HREExitException;
import hre.tools.TimeKeeper;
import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.CommonTokenStream;
import vct.antlr4.generated.LangPVLLexer;
import vct.antlr4.generated.PVLParser;
import vct.col.ast.stmt.decl.ProgramUnit;

import static hre.lang.System.*;

/**
 * Parse specified code and convert the contents to COL.
 */
public class ColPVLParser extends Parser {
    @Override
    public ProgramUnit parse(CharStream input, String fileName) {
        try {
            TimeKeeper tk = new TimeKeeper();

            LangPVLLexer lexer = new LangPVLLexer(input);
            CommonTokenStream tokenStream = new CommonTokenStream(lexer);
            PVLParser parser = new PVLParser(tokenStream);
            ErrorCounter ec = errorCounter(parser, lexer, fileName);

            PVLParser.ProgramContext tree = (PVLParser.ProgramContext) parseLLSLL(
                    ec,
                    parser,
                    () -> parser.program(),
                    false
            );
            Progress("parsing pass took %dms", tk.show());
            ec.report();
            Debug("parser got: %s", tree.toStringTree(parser));

            ProgramUnit pu = PVLtoCOL.convert(tree, fileName, tokenStream, parser);
            Progress("AST conversion pass took %dms", tk.show());
            return pu;
        } catch (HREExitException e) {
            throw e;
        } catch (Exception e) {
            DebugException(e);
            Abort("Exception %s while parsing %s", e.getClass(), fileName);
        } catch (Throwable e) {
            DebugException(e);
            Warning("Exception %s while parsing %s", e.getClass(), fileName);
            throw e;
        }
        return null;
    }
}

File: parsers/src/main/java/vct/parsers/ColPVLParser.scala
package vct.parsers

import org.antlr.v4.runtime.{CharStream, CommonTokenStream}
import vct.antlr4.generated.{LangPVLLexer, PVLParser}
import vct.col.ast.GlobalDeclaration
import vct.parsers.transform.{BlameProvider, OriginProvider, PVLToCol}

case class ColPVLParser(override val originProvider: OriginProvider, override val blameProvider: BlameProvider) extends Parser(originProvider, blameProvider) {
  override def parse[G](stream: CharStream): ParseResult[G] = {
    try {
      val lexer = new LangPVLLexer(stream)
      val tokens = new CommonTokenStream(lexer)
      originProvider.setTokenStream(tokens)
      val errors = expectedErrors(tokens, LangPVLLexer.EXPECTED_ERROR_CHANNEL, LangPVLLexer.VAL_EXPECT_ERROR_OPEN, LangPVLLexer.VAL_EXPECT_ERROR_CLOSE)
      val parser = new PVLParser(tokens)
      val tree = noErrorsOrThrow(parser, lexer, originProvider) {
        parser.program()
      }
      val decls = PVLToCol[G](originProvider, blameProvider, errors).convert(tree)
      ParseResult(decls, errors.map(_._3))
    } catch {
      case m: MatchError =>
        throw ParseMatchError(m.getMessage())
    }
  }
}